### PR TITLE
basis_spline: Fix evaluation involving `nan` values and disallow knots outside of bounds.

### PR DESCRIPTION
**Pull request by bashtage**
Use nan-functions to handle nan in input vectors as well as NA extrapolation Add test that demonstrates the issue with ZERO extrapolation

### CODE REVIEW

**Comment by bashtage** [comment]
Surpasses #219.

The test added shows a bug in `basis_spline` when using zero extrapolation.   What happens is that the selected knots can be outside of the bounds.  This means that values outside of the bounds will not be zero as expected. This cannot occur when using "na" or "clip" since these will automatically adjust the data before it hits the quantile function. 

The second commit will show one possible solution to this.

**Comment by bashtage** [comment]
I am also not sure that "clip" makes sense in that it chooses the knots from the clipped data, rather than the unclipped data.  Imagine you have some tight lower and upper bounds.  When using "na", all data points output of [lower, upper] are removed before choosing knots.  This is also what my suggested patch does in the case of "zero".  When using clipping, the quantiles are computed using a data set with repetition. This can lead to (perhaps) strange cases where multiple knots have the same values.  It also means that "na" clipping and "clip" clipping can produce different outputs aside from the replacement values in for the extrapolated values.

**Comment by bashtage** [comment]
I did see your point about extrapolation not needing to be stored - invariant keyword arguments are retained through the `stateful_transform` decorator. It is only the values which might differ from the keyword argument, e.g., lower_bound, that need to be stored in the dict.

**Comment by bashtage** [comment]
I'll wait for your thoughts on this before returning to the cubic spline since there will probably some propagation needed.

File: formulaic/transforms/basis_spline.py [hunk 1]
```diff
@@ -98,18 +98,21 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
     if df is not None and knots is not None:
         raise ValueError("You cannot specify both `df` and `knots`.")
 
+    # Ensure ndarray to avoid conversion overhead
```
Comment (matthewwardrop) [inline formulaic/transforms/basis_spline.py:101]: nit: We can probably do without this comment.

File: formulaic/transforms/basis_spline.py [hunk 2]
```diff
@@ -98,18 +98,21 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
...
+    x = numpy.asarray(x)
+
     if "lower_bound" in _state:
-        lower_bound = _state["lower_bound"]
+        lower_bound = float(_state["lower_bound"])
```
Comment (matthewwardrop) [inline formulaic/transforms/basis_spline.py:105]: nit: Let's default to trusting what is in the `_state` dictionary. These are values that have already been prepared by a previous run through this code. And, even if they were numpy floats, it doesn't actually matter in practice. (also below)

File: formulaic/transforms/basis_spline.py [hunk 3]
```diff
@@ -98,18 +98,21 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
...
     else:
         lower_bound = _state["lower_bound"] = (
-            numpy.min(x) if lower_bound is None else lower_bound
+            float(numpy.nanmin(x)) if lower_bound is None else float(lower_bound)
```
Comment (matthewwardrop) [inline formulaic/transforms/basis_spline.py:108]: nit: Take the float out of the parentheses so we can apply it just once. (also below)

File: formulaic/transforms/basis_spline.py [hunk 4]
```diff
@@ -122,10 +125,18 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
...
+        SplineExtrapolation.NA,
+        SplineExtrapolation.ZERO,
+    ):
+        valid_x = (x >= lower_bound) & (x <= upper_bound)
+        if extrapolation is SplineExtrapolation.CLIP:
+            x = numpy.clip(x, lower_bound, upper_bound)
+        elif extrapolation is SplineExtrapolation.NA:
+            x = numpy.where((x >= lower_bound) & (x <= upper_bound), x, numpy.nan)
+    else:
+        valid_x = numpy.ones_like(x, dtype=bool)
```
Comment (matthewwardrop) [inline formulaic/transforms/basis_spline.py:139]: nit: It would likely be (in most cases only marginally) cheaper to pre-compute the data degrees of freedom here, rather than building a new ones array in memory.

**Review by matthewwardrop — APPROVED** [review]
As the branch names suggests, these changes are all pretty uncontroversial (lol... seeing that made me chuckle). I left a few nitty comments, but this just hardens things up against certain edge-cases, which is fine by me! Once the nitty comments are addressed or disregarded, we can merge these in.

File: formulaic/transforms/basis_spline.py [hunk 5]
```diff
@@ -138,11 +145,18 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
...
                 )
-            knots = list(
-                numpy.quantile(x, numpy.linspace(0, 1, nknots + 2))[1:-1].ravel()
+            if valid_x.sum(0) < nknots:
+                raise ValueError(
+                    f"Insufficient valid data points to compute {nknots} knots."
+                )
+            knots = cast(
+                List[float],
+                numpy.nanquantile(x[valid_x], numpy.linspace(0, 1, nknots + 2))[
```
Comment (matthewwardrop) [inline formulaic/transforms/basis_spline.py:156]: This truncation makes sense to me too. As you noted, this ensures that the clipped and nan extrapolations are consistent where they overlap. I think this was likely just an oversight in my original implementation; since it definitely doesn't make sense that you can have knots outside of your upper and lower bounds.

File: tests/transforms/test_basis_spline.py [hunk 6]
```diff
@@ -510,3 +511,26 @@ def test_statefulness(self, data):
             ),
         ):
             basis_spline([-2, 2], extrapolation="raise", _state=state)
+
+    @pytest.mark.parametrize("extrapolation", ["na", "zero"])
+    @pytest.mark.parametrize("bounds", [[0.0, 1.0], [0.25, 0.75]])
+    def test_extrapolation_zero(self, extrapolation, bounds):
```
Comment (matthewwardrop) [inline tests/transforms/test_basis_spline.py:517]: This test case looks good to me too, thank you!

File: formulaic/transforms/basis_spline.py (hunk 2 above)
Comment (matthewwardrop) [reply formulaic/transforms/basis_spline.py:105]: Nevermind... I realised that without it you have to add ugly typing annotations anyway. Ignore this one!

File: formulaic/transforms/basis_spline.py (hunk 4 above)
Comment (matthewwardrop) [reply formulaic/transforms/basis_spline.py:139]: Actually, would it make sense to just remove this dof exception altogether? Sure, it may not make a great deal of sense for small datasets, but quantile does interpolation, and it probably stops making sense a lot earlier than when the row count is less than the number of knots... But it might be better to just output splines anyway, so that the same algorithm doesn't fail seemingly randomly on datasets with different cardinalities. Not sure if you have strong opinions about this.

File: formulaic/transforms/basis_spline.py (hunk 1 above)
Comment (bashtage) [reply formulaic/transforms/basis_spline.py:101]: Removed.

File: formulaic/transforms/basis_spline.py (hunk 4 above)
Comment (bashtage) [reply formulaic/transforms/basis_spline.py:139]: I have rewritten things so that if extrapolation is extend that there is no selection at all.  It added a bit more cyclomatic complexity, but I think it reduces temp storage and in many cases (when LB and UB are set to min() and max()) will completely eliminate any copying.

File: formulaic/transforms/basis_spline.py [hunk 7]
```diff
@@ -136,12 +150,19 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
                 raise ValueError(
                     f"Invalid value for `df`. `df` must be greater than {degree + (1 if include_intercept else 0)} [`degree` (+ 1 if `include_intercept` is `True`)]."
                 )
-            knots = list(
-                numpy.quantile(x, numpy.linspace(0, 1, nknots + 2))[1:-1].ravel()
+            if knots_x.shape[0] < nknots:
```
Comment (bashtage) [inline formulaic/transforms/basis_spline.py:153]: I think there was a comment that pointed to a different line. This could stay or go.  Could go to a minimal check that there is at least 1 data point here.

File: formulaic/transforms/basis_spline.py [hunk 8]
```diff
@@ -136,12 +150,21 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
                 raise ValueError(
                     f"Invalid value for `df`. `df` must be greater than {degree + (1 if include_intercept else 0)} [`degree` (+ 1 if `include_intercept` is `True`)]."
                 )
-            knots = list(
-                numpy.quantile(x, numpy.linspace(0, 1, nknots + 2))[1:-1].ravel()
+            if knots_x.shape[0] == 0:
```
Comment (bashtage) [inline formulaic/transforms/basis_spline.py:153]: After adding a test where the selection above removed all datapoints, so that `numpy.nanquantile([], [0, 1])` was raising an opaque error, I changed this condition to require at least 1 data point.

File: formulaic/transforms/basis_spline.py (hunk 3 above)
Comment (bashtage) [reply formulaic/transforms/basis_spline.py]: I refactored this to avoid 1 level of conditional nesting.

File: formulaic/transforms/basis_spline.py [hunk 9]
```diff
@@ -122,10 +124,22 @@ def basis_spline(  # pylint: disable=dangerous-default-value  # always replaced
...
-    if extrapolation is SplineExtrapolation.CLIP:
-        x = numpy.clip(x, lower_bound, upper_bound)
-    if extrapolation is SplineExtrapolation.NA:
-        x = numpy.where((x >= lower_bound) & (x <= upper_bound), x, numpy.nan)
+    if extrapolation in (
+        SplineExtrapolation.CLIP,
+        SplineExtrapolation.NA,
+        SplineExtrapolation.ZERO,
+    ):
+        locs = (x >= lower_bound) & (x <= upper_bound)
```
Comment (bashtage) [inline formulaic/transforms/basis_spline.py:132]: This avoids copies in most applications IME, but added some complexity to the code.

File: formulaic/transforms/basis_spline.py (hunk 8 above)
Comment (matthewwardrop) [reply formulaic/transforms/basis_spline.py:153]: Looks good. Thanks @bashtage!

**Comment by matthewwardrop** [comment]
Thanks again @bashtage !
